File: python/fastapi/routers/voice/voice_handler.py
import logging

import httpx
from fastapi import APIRouter
from fastapi import Header
from fastapi import HTTPException
from fastapi.responses import Response
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from .elevenlabs_sample import make_elevenlabs_tts_sample
from .elevenlabs_streaming_sample import make_elevenlabs_stream_tts_sample

logger = logging.getLogger(__name__)

voice_router = APIRouter()


class TTSRequest(BaseModel):
    text: str


class StreamTTSRequest(BaseModel):
    text: str


@voice_router.post("/tts/{voice}")
async def process_text_to_speech(
    voice: str,
    request: TTSRequest,
    unith_voice_x_api: str = Header(None, description="API Key for authentication"),
):
    """
    Process text-to-speech audio using Eleven Labs API for the specified voice.
    voice: The voice ID to use for TTS.
    request: The TTSRequest object containing the text to convert.
    unith_voice_x_api: API key for authentication.
    Returns a Response with audio data.
    """
    # Validate the API key
    if unith_voice_x_api != "12345678":
        raise HTTPException(status_code=401, detail="Invalid API Key")

    text = request.text
    logger.info(f"Processing TTS for voice: {voice} with text: {text}")
    try:
        audio_data: bytes = await make_elevenlabs_tts_sample(voice, request.text)
        # Return as a binary response with the correct content type
        return Response(
            content=audio_data,
            status_code=200,
            media_type="audio/wav",
            headers={"Content-Disposition": "attachment; filename=speech.wav"},
        )
    except httpx.RequestError as exc:
        logger.error(f"Request error: {exc}")
        raise HTTPException(status_code=503, detail=f"Service unavailable: {exc}")
    except Exception as exc:
        logger.error(f"Unexpected error: {exc}")
        raise HTTPException(status_code=500, detail=f"Unexpected error: {str(exc)}")


@voice_router.post("/stream-tts/{voice}")
async def stream_text_to_speech(
        voice: str,
        request: StreamTTSRequest,
        unith_voice_x_api: str = Header(None, description="API Key for authentication"),
):
    """
    Stream text-to-speech audio using Eleven Labs API streaming for the specified voice.
    voice: The voice ID to use for TTS.
    request: The TTSRequest object containing the text to convert.
    unith_voice_x_api: API key for authentication.
    Returns a StreamingResponse with audio data.
    """
    try:
        # Validate the API key
        if unith_voice_x_api != "12345678":
            raise HTTPException(status_code=401, detail="Invalid API Key")

        text = request.text
        logger.info(f"Starting streaming TTS for voice: {voice} with text: {text}")

        # Use our stream_tts generator function from tts.py
        audio_stream = make_elevenlabs_stream_tts_sample(
            voice_id=voice,
            text=request.text,
        )

        # Return streaming response
        return StreamingResponse(
            audio_stream,
            media_type="audio/wav",
            headers={"Content-Disposition": "attachment; filename=speech.wav"},
        )

    except ValueError as exc:
        logger.error(f"Validation error: {exc}")
        raise HTTPException(status_code=400, detail=str(exc))
    except httpx.RequestError as exc:
        logger.error(f"Request error: {exc}")
        raise HTTPException(status_code=503, detail=f"Service unavailable: {exc}")
    except httpx.HTTPStatusError as exc:
        logger.error(f"HTTP error: {exc}")
        status_code = exc.response.status_code if exc.response else 500
        raise HTTPException(status_code=status_code, detail=f"HTTP error: {exc}")
    except Exception as exc:
        logger.error(f"Unexpected error: {exc}")
        raise HTTPException(status_code=500, detail=f"Unexpected error: {str(exc)}")

File: nodejs/express/routers/voice/voice_handler.js
const express = require('express');
const { body, validationResult } = require('express-validator');
const {makeElevenlabsTtsSample} = require("./elevenlabs_sample");

const router = express.Router();

/**
 * @swagger
 * /tts/{voice}:
 *   post:
 *     tags:
 *       - Voice
 *     summary: Process text to speech
 *     description: Converts text to speech using the specified voice
 *     security:
 *       - apiKey: []
 *     parameters:
 *       - in: path
 *         name: voice
 *         required: true
 *         schema:
 *           type: string
 *         description: The voice to use for text-to-speech conversion
 *       - in: header
 *         name: unith-voice-x-api
 *         required: true
 *         schema:
 *           type: string
 *         description: API key for authentication
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             required:
 *               - text
 *             properties:
 *               text:
 *                 type: string
 *                 description: The text to convert to speech
 *     responses:
 *       200:
 *         description: Audio data in WAV format
 *         content:
 *           audio/wav:
 *             schema:
 *               type: string
 *               format: binary
 *       400:
 *         description: Bad request, missing or invalid parameters
 *       401:
 *         description: Unauthorized, invalid API key
 *       500:
 *         description: Server error
 */
router.post('/tts/:voice', [
  // Validate request body
  body('text').notEmpty().withMessage('Text is required'),
], async (req, res) => {
  // Check for validation errors
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ errors: errors.array() });
  }

  // Validate API key
  const apiKey = req.header('unith-voice-x-api');
  if (apiKey !== '12345678') {
    return res.status(401).json({ detail: 'Invalid API Key' });
  }

  const voice = req.params.voice;
  const text = req.body.text;

  console.log(`Processing TTS for voice: ${voice} with text: ${text}`);

  try {
    // In a real implementation, this would call a TTS service
    // For now, we'll just return an empty audio buffer
    const audioData = await makeElevenlabsTtsSample(voice, text);

    // Return as a binary response with the correct content type
    res.setHeader('Content-Type', 'audio/wav');
    res.setHeader('Content-Disposition', 'attachment; filename=speech.wav');
    return res.status(200).send(audioData);
  } catch (error) {
    console.error(`Unexpected error: ${error}`);
    return res.status(500).json({ detail: `Unexpected error: ${error.message}` });
  }
});

module.exports = router;
